core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go

```diff
@@ -162,3 +182,67 @@ type UpkeepKeyHelper[T uint32 | int64] struct {
 func (kh UpkeepKeyHelper[T]) MakeUpkeepKey(b T, id *big.Int) ocr2keepers.UpkeepKey {
 	return ocr2keepers.UpkeepKey(fmt.Sprintf("%d%s%s", b, separator, id))
 }
+
+func decodeExtensions(result *ocr2keepers.CheckResult) error {
+	if result == nil {
+		return fmt.Errorf("non-nil value expected in decoding")
+	}
+
+	rC := *result
+
+	// decode the trigger extension data if not decoded
+	switch typedExt := rC.Payload.Trigger.Extension.(type) {
+	case logprovider.LogTriggerExtension:
+		// no decoding required
+		break
+	case []byte:
+		switch getUpkeepType(result.Payload.Upkeep.ID) {
+		case logTrigger:
+			var ext logprovider.LogTriggerExtension
+
+			// in the case of a string, the value is probably still json
+			// encoded and coming from a plugin outcome
+			if err := json.Unmarshal(typedExt, &ext); err != nil {
+				return fmt.Errorf("%w: json encoded values do not match LogTriggerExtension struct: %s", err, string(typedExt))
+			}
+
+			result.Payload.Trigger.Extension = ext
+		case conditionTrigger:
+			// no special handling for conditional triggers
+			break
+		default:
+			return fmt.Errorf("unknown upkeep type")
+		}
+	case struct{}, nil:
+		// empty fallback
+		break
+	default:
+		return fmt.Errorf("unrecognized trigger extension data")
+	}
+
+	// decode the result extension data if not decoded
+	switch typedExt := result.Extension.(type) {
+	case EVMAutomationResultExtension21:
+		// no decoding required
+		break
+	case []byte:
+		var ext EVMAutomationResultExtension21
+
+		// in the case of a string, the value is probably still json
+		// encoded and coming from a plugin outcome
+		if err := json.Unmarshal(typedExt, &ext); err != nil {
+			return fmt.Errorf("%w: json encoded values do not match EVMAutomationResultExtension21 struct", err)
+		}
+
+		result.Extension = ext
+	case struct{}, nil:
+		// empty fallback
+		break
+	default:
+		return fmt.Errorf("unrecognized CheckResult extension data")
+	}
+
+	// TODO: decode upkeep config data if not decoded
```

Comment:
should this be done now? or in a follow up? If follow up I think it's a nice idea to create the ticket and then link to the ticket in the TODO comments so future devs know if the TODO was resolved.